docs/design/bundle-state.md

```diff
@@ -1,68 +1,123 @@
 
+Each UDS bundle deployed in the cluster will have its own state secret. The state secret will take the following form:
 
-### Leverage package metadata
-Add bundle information to the package metadata.
-- There isn't a specific field we could use, but we could "hijack" an existing field like `description` and add relevant bundle information there.
+```go
+type PkgStatus struct {
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Status  string `json:"status"`
+}
 
-Pros:
-- implementation would be pretty straightforward, we already do package manipulation in dev mode.
-- we would be able to leverage the `zarf package list` functionality to pull and filter on the bundle information that gets added.
+type BundleState struct {
+	Name        string      `json:"name"`
+	Version     string      `json:"version"`
+	PkgStatuses []PkgStatus `json:"packages"`
+	Status      string      `json:"status"`
+}
+```
 
-Cons:
-- we would be overloading the package metadata with bundle information, which could be confusing.
-- we compromise the integrity of the package by modifying it.
+The `BundleState` struct will be stored in the secret's `data` field as a base64 encoded JSON string.
 
-## Decision
-Use secrets to store metadata about a bundle that has been deployed to the cluster
+#### Namespace
+
+The UDS state secret will be stored in the `uds` namespace. If the `uds` namespace doesn't exist, the CLI will create it.
+
+## State Implementation
+
+Generally speaking, UDS CLI will update a bundle's state during deploy and remove operations. The bundle state will serve as a record of the bundle's deployment status and the status of each package in the bundle.
+
+### Behaviors
+
+- Bundle states will track the packages in the bundle and their statuses, if a package has been removed in future versions of a bundle, CLI will provide a mechanism to prune the cluster and state.
+- The state secret will be created during the first deployment of a bundle. If the bundle is deployed again, the state secret will be updated.
+- The state secret will be deleted when a bundle is removed.
+- If using the `--packages` flag, the CLI will only update the state for the specified packages, but the bundle secret will indicate that a package in the bundle hasn't been deployed, or in the case of removal, that a package has been removed.
+
+
+### API Design
+UDS CLI will provide a bundle state API in the form of a Go pkg called `github.com/defenseunicorns/uds-cli/src/pkg/state`. This package will provide an API for creating and interacting with state during the bundle lifecycle. Proposed public methods include:
+
+- `NewClient`: creates a new state client
+- `InitBundleState`: creates the `uds` namespace if it doesn't exist; creates a new state secret or returns an existing state secret for a previously deployed bundle
+- `GetBundleState`: retrieves the state for a given bundle
+- `GetBundlePkg`: retrieves the state for a given package inside a bundle
```

Comment:
nit: make this name conform to the others and end with `State`

Reply:
this is a really great and clear design doc.

Reply:
updated!